### ISSUE TITLE
Indent rule should ignore unknown node types

### ISSUE THREAD

**Issue by soda0289** [eslint/eslint#8440]
**Tell us about your environment**

* **ESLint Version:**
4.0.0-alpha.0
* **Node Version:**
6
* **npm Version:**
3

**What parser (default, Babel-ESLint, etc.) are you using?**
typescript

**Please show your full configuration:**
```
module.exports = {
    "parser": "typescript-eslint-parser",
    "parserOptions": {
        "ecmaVersion": 6,
        "sourceType": "module",
        "ecmaFeatures": {
            "jsx": true
        }
    },
    "rules": {
        "indent": [
            "error",
            4,
            {
                MemberExpression: 1,
                SwitchCase: 1,
                FunctionDeclaration: {
                    body: 1,
                    parameters: 1
                },
                FunctionExpression: {
                    body: 1,
                    parameters: 1
                },
                CallExpression: {
                    arguments: 1
                }
            }
        ]
    }
};

```

**What did you do? Please include the actual source code causing the issue.**

```ts
interface Foo {
    bar: string;
}
```

**What did you expect to happen?**
It should ignore unknown nodes.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/mnt/c/Users/reyad.attiyat/Workspace/eslint-test/test-interface.ts
  2:1  error  Expected indentation of 0 spaces but found 4  indent
```
It expects the interface declaration to have no indentation.

**Comment by soda0289** [comment]
@not-an-aardvark I was thinking about implement a fix for this. Would we need a whitelist of known nodes that should be indented and all others be ignored?

**Comment by not-an-aardvark** [comment]
Yes, I think that would probably be the best way to do it. Another option would be to use `espree.Syntax`, but that could cause breakage if new node types were added to espree in a minor release. For now, I suppose you could use `espree.Syntax` to get a one-time list of node types, and then hardcode that list into the file.

We would also have to make sure that indentation gets checked if a node is nested inside an ignored node. This is generally how ignored tokens work within the rule anyway, but we would need to accurately mark the right tokens as "ignored". The best way to do this might be to use `sourceCode.getNodeByRangeIndex` right before reporting a token for having incorrect indentation, and ignore the token if it's in an unknown node type. (Using `sourceCode.getNodeByRangeIndex` on every token in the file probably isn't feasible for performance reasons.)

**Comment by not-an-aardvark** [comment]
@soda0289 Are you working on this? I'm hoping we can get it done in time for the next release since it's a blocker for 4.0.0. (I can work on it if you haven't started yet -- I just want to make sure someone handles it.)

**Comment by soda0289** [comment]
@not-an-aardvark I didn't have time this weekend. I've just been trying to understand the code and find an optimal solution. I will make an attempt at a solution tonight and let you know how it goes.
